Added a footer widget for ReorderableListView

ueman opened the pull request:
## Description

This PR adds a footer widget to [ReorderableListView](https://api.flutter.dev/flutter/material/ReorderableListView-class.html). It works the same way as the existing header widget.
It also fixes the reordering in reversed list as described in https://github.com/flutter/flutter/issues/51359

## Related Issues

Fixes https://github.com/flutter/flutter/issues/21195

## Tests

I added the following tests:
- I've added two tests which are basically the same as the header tests. One for a horizontal and one for a vertical list.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.
<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

ueman committed:
Added a footer widget for ReorderableListView

ueman committed:
Added tests for ReorderableListView footer

ueman commented:
@googlebot I fixed it.

shihaohong commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -112,6 +113,19 @@ void main() {
         expect(listItems, orderedEquals(<String>['Item 2', 'Item 3', 'Item 4', 'Item 1']));
       });
 
+      testWidgets('properly reorders with a footer', (WidgetTester tester) async {
```
It seems like this test is failing presubmit checks.

shihaohong replied:
@ueman Could you explain what you changed here to get it to pass? I wasn't able to see the differences because you force-pushed, so the commit history was lost

ueman replied:
Sure. I've used the test ("properly reorders with a header") directly above it as a base for my test. 
What I've changed is that I've wrapped the footer in a SizedBox to give it the same width as normal list entry, so that I know how wide it is because the drop area is after the footer (which is the behaviour in a reversed horizontal list). Then I've also changed the footer text from "Footer Text" to just "Footer" to make sure it fits inside the SizedBox.

ueman replied:
```dart
testWidgets('properly reorders with a footer', (WidgetTester tester) async {
        await tester.pumpWidget(build(header: const Text('Footer Text'), scrollDirection: Axis.horizontal));
        expect(find.text('Footer Text'), findsOneWidget);
        expect(listItems, orderedEquals(originalListItems));
        await longPressDrag(
          tester,
          tester.getCenter(find.text('Item 1')),
          tester.getCenter(find.text('Item 4')) + const Offset(itemHeight * 2, 0.0),
        );
        await tester.pumpAndSettle();
        expect(find.text('Footer Text'), findsOneWidget);
        expect(listItems, orderedEquals(<String>['Item 2', 'Item 3', 'Item 4', 'Item 1']));
        await tester.pumpWidget(build(header: const Text('Footer Text'), scrollDirection: Axis.horizontal));
        await longPressDrag(
          tester,
          tester.getCenter(find.text('Item 4')),
          tester.getCenter(find.text('Item 3')),
        );
        expect(find.text('Footer Text'), findsOneWidget);
        expect(listItems, orderedEquals(<String>['Item 2', 'Item 4', 'Item 3', 'Item 1']));
      });
```

shihaohong replied:
I do think that there is a real issue with your implementation. The intermediary check was accidentally removed:
```
        await longPressDrag(
          tester,
          tester.getCenter(find.text('Item 1')),
          tester.getCenter(find.text('Item 4')) + const Offset(itemHeight * 2, 0.0),
        );
        await tester.pumpAndSettle();
        expect(find.text('Footer Text'), findsOneWidget);
        expect(listItems, orderedEquals(<String>['Item 2', 'Item 3', 'Item 4', 'Item 1']));
```

In a real app, the behavior is prevalent as well. You cannot drag an item from the center to the horizontal end of the list.

![reorderable_bug](https://user-images.githubusercontent.com/27032613/75373665-8ee9cf80-587f-11ea-9612-2b5becd98116.gif)

Could you please fix this issue?

ueman replied:
Yes. That could take however a few days as I'm busy.
On a side note, should the additional space after list still be there if a footer is existent? I didn't find anything about that on [https://material.io/components/lists/#types](https://material.io/components/lists/#types).

shihaohong replied:
I'm not sure what space you're referring to. Would this be space between the last item and the edges of the widget?

ueman replied:
I'm referring to the finalDropArea here:  https://github.com/flutter/flutter/blob/0b8abb4724/packages/flutter/lib/src/material/reorderable_list.dart#L544
It looks kinda weird if it's before the footer, so I left it after the footer. This also causes the effect described by you that you can't drag an item between the last item of the list and the footer. To order the item to the last position you have to drag it to the finalDropArea.
In the following screenshot I've marked finalDropArea in red.
![Simulator Screen Shot - iPhone 11 Pro Max - 2020-02-26 at 20 16 00](https://user-images.githubusercontent.com/1270149/75378920-d9496b80-58d4-11ea-9dac-4353462eea56.png)

shihaohong commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -30,7 +30,7 @@ void main() {
     }
 
-    Widget build({ Widget header, Axis scrollDirection = Axis.vertical, TextDirection textDirection = TextDirection.ltr }) {
+    Widget build({ Widget header, Widget footer, Axis scrollDirection = Axis.vertical, TextDirection textDirection = TextDirection.ltr }) {
```
We could just spread this out to make it more readable, since we have a lot of named parameters now
```suggestion
    Widget build({ 
      Widget header, 
      Widget footer, 
      Axis scrollDirection = Axis.vertical, 
      TextDirection textDirection = TextDirection.ltr,
    }) {
```

shihaohong reviewed (commented):
There seems to be a failing test, which could mean that the footer implementation might have broke something, or that the test has a problem.

ueman commented:
This should be ready to merge now. Let me know if there is anything left to do.

shihaohong commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -112,6 +118,19 @@ void main() {
 
+      testWidgets('properly reorders with a footer', (WidgetTester tester) async {
+        await tester.pumpWidget(build(footer: const Text('Footer Text')));
+        expect(find.text('Footer Text'), findsOneWidget);
```
I made a comment about this earlier, but it seems like there is a check here that is missing that would point out a bug in the implementation: 
```
        await longPressDrag(
          tester,
          tester.getCenter(find.text('Item 1')),
          tester.getCenter(find.text('Item 4')) + const Offset(itemHeight * 2, 0.0),
        );
        await tester.pumpAndSettle();
        expect(find.text('Footer'), findsOneWidget);
        expect(listItems, orderedEquals(<String>['Item 2', 'Item 3', 'Item 4', 'Item 1']));
```

In a sample app I created with a ReorderableListView, I cannot manually drag an item from any location to the end of the reorderable list view right before a footer if it exists.

ueman replied:
I've added the missing part.

ueman committed:
Fix reordering in reversed ReorderableListVIew

https://github.com/flutter/flutter/issues/51359

shihaohong commented on packages/flutter/lib/src/material/reorderable_list.dart:
```diff
@@ -581,9 +590,10 @@ class _ReorderableListContentState extends State<_ReorderableListContent> with T
         child: _buildContainerForScrollDirection(
           children: <Widget>[
-            if (widget.reverse) _wrap(finalDropArea, widget.children.length, constraints),
+            if (widget.reverse) _wrap(finalDropArea, 0, constraints),
```
Could you explain this change? I'm not clear what this does

ueman replied:
This fixes https://github.com/flutter/flutter/issues/51359.

ueman replied:
I've added a comment explaining why it must be 0.

shihaohong replied:
We should make this fix a separate pull request. I think it starts to get confusing what the goal of this PR is when you start adding fixes to non-related issues. You would also have to add a test that verifies that this behavior has been fixed so that the framework does not regress. 

Could you either revert this and create a new PR for this fix, or add a test here and change the PR title and description?

ueman replied:
I've reverted the unrelated changed and made a new PR for them.

ueman committed:
ReorderableListView

- Fixed the behaviour of the finalDropArea if a footer is present
- wrote a test to make sure the an item which gets dragged to the start of a reversed list is also ordered to the start
- improved code documentation

ueman committed:
Better test documentation and implementation

shihaohong commented on packages/flutter/lib/src/material/reorderable_list.dart:
```diff
@@ -581,10 +590,25 @@ class _ReorderableListContentState extends State<_ReorderableListContent> with T
+            // We don't need a final drop area when a footer is present
+            // as the footer should provide enough space to drag a list item
+            // to the last position of the list, which is above the footer.
+            if (!widget.reverse && widget.footer == null) _wrap(finalDropArea, widget.children.length, constraints),
```
I just tested this behavior with a footer that is a `SizedBox` with width set to something small like 1-5, and this doesn't work unless you drag your item into the vicinity of the box. If were to drag it down to the very bottom of the screen quickly, it doesn't seem to reorder correctly. Removing the footer restores this behavior. You will need to make the drop area the same size as it were before if the footer is smaller than `finalDropArea`, or use the footer's height as the final drop height otherwise.

shihaohong commented on packages/flutter/lib/src/material/reorderable_list.dart:
```diff
@@ -581,10 +590,25 @@ class _ReorderableListContentState extends State<_ReorderableListContent> with T
+              // We need to wrap the widget.footer so we can reorder an item between
+              // the last item of the list and the footer itself
+              _wrap(SizedBox(
+                  key: Key('footer'),
```
Why is adding the key here necessary?

Ahmadre replied:
I added footer locally and I encountered, that the footer also has to have key. Don't know why but it seems to ReorderableListView checks for it.

shihaohong reviewed (changes requested):
I have some extra feedback -- I think you should solve the two issues your PR attempts to fix in separate PRs to make reviewing these changes easy. `ReorderableListView`'s mechanics are non-trivial and making these changes can be confusing to future readers who may not have a lot of context for how it works. We also want to make sure we don't cause any unexpected breaking changes.

ueman committed:
Revert "Fix reordering in reversed ReorderableListVIew"

This reverts commit 50572d1c9872b65a89a2b85007e06b5b4ac28102.

# Conflicts:
#	packages/flutter/lib/src/material/reorderable_list.dart

ueman committed:
Removed fixed reodering in reverse to keep this PR focused

ueman committed:
A list element can be reordered between the footer and the last list item

ueman committed:
Align footer right in horizontal mode

ueman committed:
footer does not need to be wrapped inside a SizedBox anymore because the footer is now constrained to a size by the ReorderableListView

ueman committed:
complete test for vertical ReorderableListView

ueman committed:
Use correct alignment

ueman committed:
Remove trailing whitespace

shihaohong commented:
I had a discussion with some other members of the Flutter team about adding a footer -- Could you describe a use-case where this might be useful? We were wondering why it wouldn't be possible to say, use a `Column` widget with the ReorderableListView followed by whatever widget you want.

For example, wouldn't some thing like this achieve a similar effect:
```dart
Column(
  children: <Widget>[
    ReorderableListView(...),
    Text('Hello World'),
  ],
)
```

Ahmadre commented:
> I had a discussion with some other members of the Flutter team about adding a footer -- Could you describe a use-case where this might be useful? We were wondering why it wouldn't be possible to say, use a `Column` widget with the ReorderableListView followed by whatever widget you want.
> 
> For example, wouldn't some thing like this achieve a similar effect:
> 
> ```dart
> Column(
>   children: <Widget>[
>     ReorderableListView(...),
>     Text('Hello World'),
>   ],
> )
> ```

I can give you exactly a Use-Case: I got a costumer who wants a reorderable list of items in our EndDrawer. Now we have also "Add Page" Buttons underneeth. And when I wrap it with a Column it goes to Rendering Errors! And I changed Reorderablelistview to have a footer and it works even better then wrapping it outside with a column! Here's my screen:
<img width="435" alt="Bildschirmfoto 2020-03-14 um 20 03 49" src="https://user-images.githubusercontent.com/18512224/76688637-eda89a80-662e-11ea-9e71-30d8ddabc971.png">

ueman commented:
@Ahmadre You can do something like (the footer will always be visible at the bottom of the screen). You also can't get both of them to scroll together. If tried this example https://api.flutter.dev/flutter/widgets/SingleChildScrollView-class.html#expanding-content-to-fit-the-viewport but it doesn't work
```dart
    Column(
        mainAxisSize: MainAxisSize.min,
        children: <Widget>[
          Flexible(
            fit: FlexFit.loose,
            child: ReorderableListView(...),
          ),
          Flexible(
            fit: FlexFit.loose,
            child: FlutterLogo(),
          ),
        ]
```
However there is still this ugly whitespace at the end of the `ReorderableListView` if you scroll to the bottom: 
![Simulator Screen Shot - iPhone 11 Pro Max - 2020-03-15 at 10 39 01](https://user-images.githubusercontent.com/1270149/76699014-356c0800-66a9-11ea-9d82-5606c9cd9013.png)

The whitespace between the list and the footer and because you can't get a real footer is why I made this PR.

Ahmadre commented:
@uemam: I already tried that and that solution also doesn't work, because the examples here that are presented are too easy and plain. ReaI projects aren't like this. We got a dynamic list that can grow and shrink. And if you do that like you mention your list won't adapt and items will get cut off when adding or removing items. For a dynamic list like ours it only works inside the widget itself with footer.

shihaohong commented on packages/flutter/lib/src/material/reorderable_list.dart:
```diff
@@ -584,7 +597,29 @@ class _ReorderableListContentState extends State<_ReorderableListContent> with T
+              // the last item of the list and the footer itself. If the footer
+              // is used instead of the finalDropArea it needs its own key,
+              // because _wrap() requires it.
+              _wrap(ConstrainedBox(
```
I don't quite understand what is going on here. How come the header requires the `ConstrainedBox` w/ `BoxCosntraints`, as well as the `Align` widget, whereas the header has none of these things?

I would expect that the header and footer should be rather symmetric in terms of implementation since they're basically the same thing, but on opposite ends of the ReorderableListView

shihaohong commented on packages/flutter/lib/src/material/reorderable_list.dart:
```diff
@@ -584,7 +597,29 @@ class _ReorderableListContentState extends State<_ReorderableListContent> with T
+                widget.children.length,
+                constraints,
+              ),
+            // We don't need a final drop area when a footer is present
```
This isn't necessarily true and depends on the size of the footer widget (I think I mentioned this before). We should make sure that there is some heuristic or logic to figure out large the final drop area should be if there is a footer.

Otherwise, we have to at least be very clear to users that the footer has to be big enough to drag a list item, but I feel that this is could easily be a source of confusion

shihaohong commented:
Hi @ueman, do you think you will have time to come back to this? Sorry about the delay in checking back in!

ueman commented:
Yeah, I think that should be doable.

shihaohong commented:
Hi @ueman, I'll close out this PR for now since there hasn't been any progress so far. This helps Flutter contributors to properly track open PRs that are a little more active. 

This shouldn't impact this PR's progress at all since it's pretty painless to reopen the PR (or create a new one) once there's progress with the work here. I'll revisit this feature once there's an update :)

DebugTheCode commented:
Any progress on this?
